src/storage/swap.cc

```diff
@@ -16,35 +18,59 @@ std::shared_ptr<Swap> Swap::_GetSharedRef() {
 
 Swap::Swap(){
   std::cout << "Initialize Swap" <<std::endl;
-  mhistory_ = MemHistory::_GetSharedRef();
+  memory_history_ = MemHistory::_GetSharedRef();
+  memory_manager_ = MemoryManager::_GetSharedRef();
+  swap_lock_ = PTHREAD_RWLOCK_INITIALIZER;
+  for (int i = 0; i < NUMBER_OF_GPU; ++i){
+    locks_[i] = PTHREAD_RWLOCK_INITIALIZER;
+    free_memory_.push_back(0);
+  }
 }
 
 Swap::~Swap(){
   std::cout << "Destroy Swap" <<std::endl;
 }
 
 void Swap::SwapOut(unsigned required_memory, int device){
-  // Not implemented  
+  UpdateFree(device);
+  if (free_memory_[device] > required_memory) {
+    return;
+  }
+  while (free_memory_[device] < required_memory) {
+    handle_id_t victim = 
+      memory_history_->DecideVictim(swappable_handles_[device], device);
+    SwapInfo *target = swap_info_[victim];
+    if(target->cpu_address == nullptr) {
+      target->cpu_address = new char[int(target->size)];
+    }
+    CHECK(target->swapped_in);
+    CHECK(target->dptr != nullptr);
+    target->swapped_in = false;
+    swappable_handles_[device].erase(victim);
+    memory_manager_->Memcpy(device, target->cpu_address, target->dptr,
+        target->size, cudaMemcpyDeviceToHost);
+    memory_manager_->Free(target->dptr, device);
+    UpdateFree(device);
+  }
 }
 
 void Swap::SwapIn(SwapInfo *info){
-  // Not implemented
-}
-
-// TODO(sotskin) compatibility for MKLMEM
-void* Swap::GetAddr(handle_id_t handle_id, size_t size){
-  pthread_rwlock_rdlock(&swap_lock_);
-  auto info = swap_info_.at(handle_id);
-  if (info->device_id != -1) {
-    mhistory_->PutRecord(handle_id, info->device_id, MemHistory::DEL_ADDR, size);
+  CHECK(!info->swapped_in);
+  CHECK(info->cpu_address != nullptr);
+  SwapOut(info->size, info->device_id);
+  e = memory_manager_->Malloc(&(info->dptr), info->size);
```

Comment:
So, our memory_manager_ returns cudaSuccess too?

Reply:
For now the memory_manager_ returns cudaSuccess/Error, we think we will replace it with our own ones later